google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonToProtoMessage.java

```diff
@@ -0,0 +1,322 @@
+          } else if (val instanceof Long) {
+            protoMsg.setField(fieldDescriptor, new Long((Long) val));
+          } else {
+            throw new JSONException("");
+          }
+          break;
+        case INT32:
+          val = json.get(exactJsonKeyName);
+          if (val instanceof Integer) {
+            protoMsg.setField(fieldDescriptor, new Integer((Integer) val));
+          } else {
+            throw new JSONException("");
+          }
+          break;
+        case STRING:
+          protoMsg.setField(fieldDescriptor, json.getString(exactJsonKeyName));
+          break;
+        case DOUBLE:
+          val = json.get(exactJsonKeyName);
+          if (val instanceof Double) {
+            protoMsg.setField(fieldDescriptor, new Double((double) val));
+          } else if (val instanceof Float) {
+            protoMsg.setField(fieldDescriptor, new Double((float) val));
+          } else {
+            throw new JSONException("");
+          }
+          break;
+        case MESSAGE:
+          Message.Builder message = protoMsg.newBuilderForField(fieldDescriptor);
+          protoMsg.setField(
+              fieldDescriptor,
+              convertJsonToProtoMessageImpl(
+                  fieldDescriptor.getMessageType(),
+                  json.getJSONObject(exactJsonKeyName),
+                  currentScope,
+                  /*topLevel =*/ false,
+                  allowUnknownFields));
+          break;
+      }
+    } catch (JSONException e) {
+      throw new IllegalArgumentException(
+          String.format(
+              "JSONObject does not have a %s field at %s.",
+              FieldTypeToDebugMessage.get(fieldDescriptor.getType()), currentScope));
+    }
+  }
+
+  /**
+   * Fills a repeated protoField with the json data.
+   *
+   * @param protoMsg The protocol buffer message being constructed
+   * @param fieldDescriptor
+   * @param json If root level has no matching fields, throws exception.
+   * @param exactJsonKeyName Actual key name in JSONObject instead of lowercased version
+   * @param currentScope Debugging purposes
+   * @param allowUnknownFields Ignores unknown JSON fields.
+   * @throws IllegalArgumentException when JSON data is not compatible with proto descriptor.
+   */
+  private static void fillRepeatedField(
+      DynamicMessage.Builder protoMsg,
+      FieldDescriptor fieldDescriptor,
+      JSONObject json,
+      String exactJsonKeyName,
+      String currentScope,
+      boolean allowUnknownFields)
+      throws IllegalArgumentException {
+
+    JSONArray jsonArray;
+    try {
+      jsonArray = json.getJSONArray(exactJsonKeyName);
+    } catch (JSONException e) {
+      throw new IllegalArgumentException(
+          "JSONObject does not have a array field at " + currentScope + ".");
+    }
+    java.lang.Object val;
+    int index = -1;
+    try {
+      switch (fieldDescriptor.getType()) {
+        case BOOL:
+          for (int i = 0; i < jsonArray.length(); i++) {
```

Comment:
can we move for loop outside of switch? so that we don't need these for loops?

like :
for (int i = 0; i < jsonArray.length(); i++) {
  switch (fieldDescriptor.getType()) {
  }
}

Reply:
Yup, good catch!